File: p34/p34_taguchi.go
// go fmt && go run p34_taguchi.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var file, _ = os.Open(`./p34.in`)
var sc = bufio.NewScanner(file)

func nextLine() string {
	sc.Scan()
	return sc.Text()
}

var n, _ = strconv.Atoi(nextLine())
var sa = strings.Split(nextLine(), ",")
var a = sliceAtoi(sa)

func sliceAtoi(sa []string) []int {
	si := make([]int, 0, len(sa))
	for _, s := range sa {
		i, _ := strconv.Atoi(s)
		si = append(si, i)
	}
	return si
}

var k, _ = strconv.Atoi(nextLine())

func check(i int, sum int) bool {
	if i == n {
		return k == sum
	} else if check(i+1, sum+a[i]) {
		return true
	} else if check(i+1, sum) {
		return true
	}
	return false
}

func main() {
	if check(0, 0) {
		fmt.Println("Yes")
	} else {
		fmt.Println("No")
	}
}

File: p34/p34_taguchi.rb
array = []
File.open('./p34.in') { |file| file.each_line { |line| array << line.chomp } }
@a = array[1].split(',').map(&:to_i)
@k = array[2].to_i

def check(i, sum)
  if i == @a.count then @k == sum
  elsif check(i + 1, sum + @a[i]) then true
  elsif check(i + 1, sum) then true
  else false
  end
end

check(0, 0) ? p('Yes') : p('No')
